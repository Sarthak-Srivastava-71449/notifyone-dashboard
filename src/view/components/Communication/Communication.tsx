import React, { useEffect, useState } from 'react';
import { Layout } from 'antd';
import { useLocation, useNavigate, useParams } from 'react-router-dom';

import { COMMUNICATION_TYPE } from 'src/common/constants';

import { CommunicationTemplate } from '../CommunicationTemplate';
import { commonEventDetailType } from '../types';
import { useDispatch, useSelector } from 'react-redux';
import { RootState } from 'src/store';
import { getWhatsAppEvents } from 'src/store/selectors/whatsAppEvents';
import { fetchCurrentEmailEvent } from 'src/store/actions/emailEvents';
import { getAccessToken } from 'src/store/selectors/accessToken';
import { Spinner } from '../Spinner';
import { fetchCurrentSmsEvent } from 'src/store/actions/smsEvents';
import { fetchCurrentPushEvent } from 'src/store/actions/pushNotificationEvents';
import { fetchCurrentWhatsappEvent } from 'src/store/actions/whatsappEvents';

const { Content } = Layout;

const Communication: React.FC = () => {
  const location = useLocation();
  const dispatch = useDispatch();
  const navigate = useNavigate();
  const { id } = useParams();
  const currentEvent = useSelector(
    (state: RootState) => state.currentEvent.current_event
  );
  const whatsAppEvents = useSelector((state: RootState) =>
    getWhatsAppEvents(state)
  );
  const accessToken = useSelector((state: RootState) => getAccessToken(state));

  const currentEventLoading = useSelector(
    (state: RootState) => state.currentEvent.loading
  );

  const [eventDetails, setEventDetails] = useState<commonEventDetailType>({
    app_name: '',
    event_name: '',
    event_type: '',
    id: '',
    triggers_limit: '',
  });

  const shouldDisable = !!currentEvent;

  useEffect(() => {
    if (!currentEvent) {
      if (
        location.pathname.startsWith('/communication/template/email/') &&
        id
      ) {
        dispatch(fetchCurrentEmailEvent(accessToken, parseInt(id)));
      } else if (
        location.pathname.startsWith('/communication/template/sms/') &&
        id
      ) {
        dispatch(fetchCurrentSmsEvent(accessToken, parseInt(id)));
      } else if (
        location.pathname.startsWith('/communication/template/transaction/') &&
        id
      ) {
        dispatch(fetchCurrentPushEvent(accessToken, parseInt(id)));
      } else if (
        location.pathname.startsWith('/communication/template/whatsapp/') &&
        id
      ) {
        dispatch(fetchCurrentWhatsappEvent(accessToken, parseInt(id)));
      }
    }
  }, [location.pathname]);

  useEffect(() => {
    if (currentEvent && currentEvent !== null) {
      const record: any = currentEvent;

      const eventProperties = {
        actions: record?.actions,
        app_name: record?.app_name,
        event_name: record?.event_name,
        event_type: record?.event_type,
        id: record?.id,
        triggers_limit: record?.triggers_limit,
        event_id: record?.event_id,
        data: record?.data,
      };

      if (record.event_type === COMMUNICATION_TYPE.SMS) {
        return setEventDetails({
          ...eventProperties,
          event_text: record?.content,
        });
      }

      if (
        record.event_type === COMMUNICATION_TYPE.Transactional_Push_notification
      ) {
        return setEventDetails({
          ...eventProperties,
          event_text: record?.body,
          image: record?.image,
          target: record?.target,
          title: record?.title,
          updated: record?.updated,
          updated_by: record?.updated_by,
        });
      }

      if (record.event_type == COMMUNICATION_TYPE.Whatsapp) {
        return setEventDetails({
          ...eventProperties,
          event_text: record?.name,
        });
      }

      if (record.event_type == COMMUNICATION_TYPE.Email) {
        return setEventDetails({
          ...eventProperties,
          event_text: record?.content,
          description: record?.description,
          includes: record?.includes,
          name: record?.name,
          subject: record?.subject,
          updated_by: record?.updated_by,
        });
      }
    }
  }, [currentEvent, currentEventLoading]);

  return (
    <Layout
      hasSider
      className="max-h-[calc(100vh-50px-50px)] overflow-y-scroll"
    >
      {currentEventLoading ? (
        <Spinner
          className="h-[calc(100vh-48px-50px)]"
          loading={currentEventLoading}
        />
      ) : (
        <Content>
          <CommunicationTemplate
            eventDetails={eventDetails}
            setEventDetails={setEventDetails}
            shouldDisable={shouldDisable}
          />
        </Content>
      )}
    </Layout>
  );
};

export default Communication;
