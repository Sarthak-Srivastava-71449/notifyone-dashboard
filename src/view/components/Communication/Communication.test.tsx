import React from 'react';
import { render, screen } from '@testing-library/react';
import * as routeData from 'react-router';

import {
  mockSmsEventDetails,
  mockEmailEventDetails,
  mockPushNotificationEventDetails,
  mockWhatsappEventDetails,
  withProvider,
} from 'src/common/renderWithProvider';

import Communication from './Communication';

describe('Communication', () => {
  test('should display Communication screen', async () => {
    const { asFragment } = render(withProvider(<Communication />));
    expect(asFragment()).toMatchSnapshot();
  });

  test('should display sms event details when sms event is provided in location.state', async () => {
    jest.spyOn(routeData, 'useLocation').mockReturnValue({
      state: { record: mockSmsEventDetails },
      key: '',
      pathname: '',
      search: '',
      hash: '',
    });

    render(withProvider(<Communication />));

    const communicationTypeSelectOption = screen.getByText('SMS');

    expect(communicationTypeSelectOption).toBeInTheDocument();
  });

  test('should display email event details when email event is provided in location.state', async () => {
    jest.spyOn(routeData, 'useLocation').mockReturnValue({
      state: { record: mockEmailEventDetails },
      key: '',
      pathname: '',
      search: '',
      hash: '',
    });

    render(withProvider(<Communication />));

    const communicationTypeSelectOption = screen.getByText('Email');

    expect(communicationTypeSelectOption).toBeInTheDocument();
  });

  test('should display whatsapp event details when whatsapp event is provided in location.state', async () => {
    jest.spyOn(routeData, 'useLocation').mockReturnValue({
      state: { record: mockWhatsappEventDetails },
      key: '',
      pathname: '',
      search: '',
      hash: '',
    });

    render(withProvider(<Communication />));

    const communicationTypeSelectOption = screen.getByText('Whatsapp');

    expect(communicationTypeSelectOption).toBeInTheDocument();
  });

  test('should display pushNotification event details when pushNotification event is provided in location.state', async () => {
    jest.spyOn(routeData, 'useLocation').mockReturnValue({
      state: { record: mockPushNotificationEventDetails },
      key: '',
      pathname: '',
      search: '',
      hash: '',
    });

    render(withProvider(<Communication />));

    const communicationTypeSelectOption = screen.getByText(
      'Transactional Push Notification'
    );

    expect(communicationTypeSelectOption).toBeInTheDocument();
  });

  // test('should update sms event details when sms event is provided and setEventDetails is called', async () => {
  //   jest.spyOn(routeData, 'useLocation').mockReturnValue({
  //     state: { record: mockSmsEventDetails },
  //     key: '',
  //     pathname: '',
  //     search: '',
  //     hash: '',
  //   });

  //   render(withProvider(<Communication />));

  //   const communicationTypeSelectOption = screen.getByText('SMS');

  //   expect(communicationTypeSelectOption).toBeInTheDocument();
  // });
});
