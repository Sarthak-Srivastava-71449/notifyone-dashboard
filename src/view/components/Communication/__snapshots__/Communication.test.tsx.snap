// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Communication should display Communication screen 1`] = `
<DocumentFragment>
  <section
    class="ant-layout ant-layout-has-sider max-h-[calc(100vh-50px-50px)] overflow-y-scroll"
  >
    <main
      class="ant-layout-content"
    >
      <div
        class="px-10 py-10"
      >
        <div
          class="ant-row"
          style="margin: -5px -8px -5px -8px;"
        >
          <div
            class="ant-col ant-col-xs-12 ant-col-sm-12 ant-col-md-4"
            style="padding: 5px 8px 5px 8px;"
          >
            <span
              class="required"
            >
              Type
            </span>
          </div>
          <div
            class="ant-col ant-col-xs-12 ant-col-sm-12 ant-col-md-8"
            style="padding: 5px 8px 5px 8px;"
          >
            <div
              class="ant-select ant-select-single ant-select-show-arrow"
              style="min-width: 100%; max-width: 100%;"
            >
              <div
                class="ant-select-selector"
              >
                <span
                  class="ant-select-selection-search"
                >
                  <input
                    aria-activedescendant="rc_select_TEST_OR_SSR_list_0"
                    aria-autocomplete="list"
                    aria-controls="rc_select_TEST_OR_SSR_list"
                    aria-haspopup="listbox"
                    aria-owns="rc_select_TEST_OR_SSR_list"
                    autocomplete="off"
                    class="ant-select-selection-search-input"
                    id="rc_select_TEST_OR_SSR"
                    readonly=""
                    role="combobox"
                    style="opacity: 0;"
                    type="search"
                    unselectable="on"
                    value=""
                  />
                </span>
                <span
                  class="ant-select-selection-item"
                  title="SMS"
                >
                  SMS
                </span>
              </div>
              <span
                aria-hidden="true"
                class="ant-select-arrow"
                style="user-select: none;"
                unselectable="on"
              >
                <span
                  aria-label="down"
                  class="anticon anticon-down ant-select-suffix"
                  role="img"
                >
                  <svg
                    aria-hidden="true"
                    data-icon="down"
                    fill="currentColor"
                    focusable="false"
                    height="1em"
                    viewBox="64 64 896 896"
                    width="1em"
                  >
                    <path
                      d="M884 256h-75c-5.1 0-9.9 2.5-12.9 6.6L512 654.2 227.9 262.6c-3-4.1-7.8-6.6-12.9-6.6h-75c-6.5 0-10.3 7.4-6.5 12.7l352.6 486.1c12.8 17.6 39 17.6 51.7 0l352.6-486.1c3.9-5.3.1-12.7-6.4-12.7z"
                    />
                  </svg>
                </span>
              </span>
            </div>
          </div>
          <div
            class="ant-col ant-col-xs-12 ant-col-sm-12 ant-col-md-4"
            style="padding: 5px 8px 5px 8px;"
          >
            <span
              class="required"
            >
              Event name
            </span>
          </div>
          <div
            class="ant-col ant-col-xs-12 ant-col-sm-12 ant-col-md-8"
            style="padding: 5px 8px 5px 8px;"
          >
            <input
              class="ant-input"
              maxlength="30"
              placeholder="Enter event name"
              type="text"
              value=""
            />
          </div>
          <div
            class="ant-col ant-col-xs-12 ant-col-sm-12 ant-col-md-4"
            style="padding: 5px 8px 5px 8px;"
          >
            <span
              class="required"
            >
              Application Name
            </span>
          </div>
          <div
            class="ant-col ant-col-xs-12 ant-col-sm-12 ant-col-md-8"
            style="padding: 5px 8px 5px 8px;"
          >
            <input
              class="ant-input"
              maxlength="30"
              placeholder="Enter application name"
              type="text"
              value=""
            />
          </div>
          <div
            class="ant-col ant-col-xs-12 ant-col-sm-12 ant-col-md-4"
            style="padding: 5px 8px 5px 8px;"
          >
            <span
              class="required"
            >
              Enabled
            </span>
          </div>
          <div
            class="ant-col ant-col-xs-12 ant-col-sm-12 ant-col-md-8"
            style="padding: 5px 8px 5px 8px;"
          >
            <button
              aria-checked="false"
              class="ant-switch bg-app-theme"
              role="switch"
              type="button"
            >
              <div
                class="ant-switch-handle"
              />
              <span
                class="ant-switch-inner"
              >
                Off
              </span>
            </button>
          </div>
          <div
            class="ant-col ant-col-xs-12 ant-col-sm-12 ant-col-md-4"
            style="padding: 5px 8px 5px 8px;"
          >
            <span
              class="required"
            >
              Trigger limit
            </span>
          </div>
          <div
            class="ant-col ant-col-xs-12 ant-col-sm-12 ant-col-md-8"
            style="padding: 5px 8px 5px 8px;"
          >
            <input
              class="ant-input"
              maxlength="30"
              pattern="[+-]?\\\\d+(?:[.,]\\\\d+)?"
              placeholder="Enter trigger limit"
              type="number"
              value=""
            />
          </div>
          <div
            class="ant-col ant-col-xs-12 ant-col-sm-12 ant-col-md-4"
            style="padding: 5px 8px 5px 8px;"
          >
            Select include
          </div>
          <div
            class="ant-col ant-col-xs-12 ant-col-sm-12 ant-col-md-8"
            style="padding: 5px 8px 5px 8px;"
          >
            <div
              class="ant-select ant-select-single ant-select-show-arrow"
              style="min-width: 100%; max-width: 100%;"
            >
              <div
                class="ant-select-selector"
              >
                <span
                  class="ant-select-selection-search"
                >
                  <input
                    aria-activedescendant="rc_select_TEST_OR_SSR_list_0"
                    aria-autocomplete="list"
                    aria-controls="rc_select_TEST_OR_SSR_list"
                    aria-haspopup="listbox"
                    aria-owns="rc_select_TEST_OR_SSR_list"
                    autocomplete="off"
                    class="ant-select-selection-search-input"
                    id="rc_select_TEST_OR_SSR"
                    readonly=""
                    role="combobox"
                    style="opacity: 0;"
                    type="search"
                    unselectable="on"
                    value=""
                  />
                </span>
                <span
                  class="ant-select-selection-item"
                  title="Select include"
                >
                  Select include
                </span>
              </div>
              <span
                aria-hidden="true"
                class="ant-select-arrow"
                style="user-select: none;"
                unselectable="on"
              >
                <span
                  aria-label="down"
                  class="anticon anticon-down ant-select-suffix"
                  role="img"
                >
                  <svg
                    aria-hidden="true"
                    data-icon="down"
                    fill="currentColor"
                    focusable="false"
                    height="1em"
                    viewBox="64 64 896 896"
                    width="1em"
                  >
                    <path
                      d="M884 256h-75c-5.1 0-9.9 2.5-12.9 6.6L512 654.2 227.9 262.6c-3-4.1-7.8-6.6-12.9-6.6h-75c-6.5 0-10.3 7.4-6.5 12.7l352.6 486.1c12.8 17.6 39 17.6 51.7 0l352.6-486.1c3.9-5.3.1-12.7-6.4-12.7z"
                    />
                  </svg>
                </span>
              </span>
            </div>
          </div>
        </div>
        <div
          class="ant-row mt-4"
        >
          <div
            class="ant-col ant-col-4"
          >
            <span
              class="required"
            >
              Event text
            </span>
          </div>
        </div>
        <div
          class="ant-row mt-4"
        >
          <div
            class="ant-col ant-col-24"
          >
            <div
              class="ant-input-textarea ant-input-textarea-show-count"
              data-count="0"
            >
              <textarea
                class="ant-input"
                placeholder="Event text / content"
                style="resize: none; height: -40px; min-height: -16px; max-height: -40px;"
              />
            </div>
          </div>
        </div>
        <div
          class="ant-row ant-row-end pt-8"
        >
          <div
            class="ant-col"
          >
            <button
              class="button primary-button min-w-[120px]"
              data-testid="preview-event-button"
            >
              Preview
            </button>
          </div>
        </div>
      </div>
    </main>
  </section>
</DocumentFragment>
`;
